src/rgw/rgw_dedup.h

```diff
@@ -0,0 +1,232 @@
+#pragma once
+#include "common/dout.h"
+#include "rgw_common.h"
+#include "rgw_dedup_utils.h"
+#include "rgw_dedup_table.h"
+#include "rgw_dedup_cluster.h"
+#include "rgw_realm_reloader.h"
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <iostream>
+#include <ostream>
+
+namespace rgw::dedup {
+  struct dedup_epoch_t;
+  struct control_t {
+    control_t() {
+      reset();
+    }
+    void reset();
+    inline bool local_urgent_req() const {
+      return (shutdown_req || local_pause_req);
+    }
+    inline bool should_stop() const {
+      return (shutdown_req || remote_abort_req);
+    }
+    inline bool should_pause() const {
+      return (local_pause_req || remote_pause_req);
+    }
+
+    // allow to start/pasue/resume/stop execution
+    dedup_req_type_t dedup_type = dedup_req_type_t::DEDUP_TYPE_NONE;
+    bool started            = false;
+    bool dedup_exec         = false;
+    bool shutdown_req       = false;
+    bool shutdown_done      = false;
+    bool local_pause_req    = false;
+    bool local_paused       = false;
+    bool remote_abort_req   = false;
+    bool remote_aborted     = false;
+    bool remote_pause_req   = false;
+    bool remote_paused      = false;
+    bool remote_restart_req = false;
+  };
+  std::ostream& operator<<(std::ostream &out, const control_t &ctl);
+  void encode(const control_t& ctl, ceph::bufferlist& bl);
+  void decode(control_t& ctl, ceph::bufferlist::const_iterator& bl);
+
+  class disk_block_seq_t;
+  struct disk_record_t;
+  struct key_t;
+  //Interval between each execution of the script is set to 5 seconds
+  static inline constexpr int INIT_EXECUTE_INTERVAL = 5;
+  class Background : public RGWRealmReloader::Pauser {
+    class DedupWatcher : public librados::WatchCtx2, public DoutPrefixProvider {
+      Background* const parent;
+    public:
+      DedupWatcher(Background* _parent) : parent(_parent) {}
+      ~DedupWatcher() override = default;
+      void handle_notify(uint64_t notify_id, uint64_t cookie,
+                         uint64_t notifier_id, bufferlist& bl) override;
+      void handle_error(uint64_t cookie, int err) override;
+
+      // DoutPrefixProvider iterface
```

Comment:
do we really need the watcher to implements the `DoutPrefixProvider` API?
can we use the parent background for that?